### PR TITLE
Notify on javascript exceptions

### PR DESCRIPTION
**Pull request by benrudolph**
Think it's time we log some of our JS errors. Probably will create quite a bit of noise at the beginning, but probably worth fixing the issues. `onerror` doesn't catch all the errors that happen, but does catch most. Also older browsers don't record the stack trace which is a bummer (new versions of chrome, firefox and opera should be fine though)

@millerdev

### CODE REVIEW

**Comment by benrudolph** [comment]
fyi @TylerSheffels

**Comment by benrudolph** [comment]
tests are failing b/c it's dependent on this: https://github.com/dimagi/dimagi-utils/pull/270

**Comment by TylerSheffels** [comment]
:+1:

**Comment by czue** [comment]
if that's true you can include the submodule update in the PR and it should fix it

**Comment by czue** [comment]
I imagine we'll find out pretty quickly whether we need to throttle this somehow. i'm ok merging but we should expect that if we start getting tons of email immediately we'll have to do a second hotfix deploy to address it

**Comment by czue** [comment]
my comment isn't blocking but will let someone else merge

**Comment by benrudolph** [comment]
Fair, i'll look into a way we can do that. perhaps using redis or logic from soft_assert

File: corehq/apps/hqwebapp/views.py [hunk 1]
```diff
@@ -395,6 +395,19 @@ def debug_notify(request):
...
+@require_POST
+def jserror(request):
+    notify_js_exception(
+        request,
+        message=request.POST.get('message', None),
+        stack=request.POST.get('stack', None),
+        line=request.POST.get('line', None),
+        filename=request.POST.get('filename', None),
+    )
+    return HttpResponse("JS error has been logged")
```
Comment (millerdev) [inline corehq/apps/hqwebapp/views.py]: Very minor, but might want to consider sending the most basic response possible here. I'm thinking something along the lines of 200 OK with no content. Anyone who hits this handler knows they're submitting a JS error, and reducing bandwidth overhead in any way possible seems ideal.

Comment (benrudolph) [inline corehq/apps/hqwebapp/views.py]: gotcha, i'll send an empty response

**Comment by benrudolph** [comment]
@czue, implemented exponential backoff using a lot of @dannyroberts code. I think we should merge the two, but thought it was out of scope for this PR

File: corehq/util/cache_utils.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        return x > 0 and (x & (x - 1) == 0)
+
+    @classmethod
+    def increment(cls, key):
+        cache_key = cls._get_cache_key(key)
+        try:
+            return cache.incr(cache_key)
+        except ValueError:
+            cache.set(cache_key, 1)
+            return 1
```
Comment (dannyroberts) [inline corehq/util/cache_utils.py]: Looks like some of this is copied from soft assert? I think this is a great abstraction. Next step is to get soft_assert using this code?

Comment (dannyroberts) [inline corehq/util/cache_utils.py]: @benrudolph Just saw you're comment; agree with you. I think just put this piece of cleanup in your shortish-term backlog and no need to block merge

File: corehq/apps/hqwebapp/views.py [hunk 3]
```diff
@@ -395,6 +396,24 @@ def debug_notify(request):
...
+
+    ExponentialBackoff.increment(stack)
+    if not ExponentialBackoff.should_backoff(stack):
+        notify_js_exception(
+            request,
+            message=request.POST.get('message', None),
+            stack=stack,
+            line=request.POST.get('line', None),
+            filename=request.POST.get('filename', None),
+        )
```
Comment (dannyroberts) [inline corehq/apps/hqwebapp/views.py]: I think this is fine, other option is to funnel to sentry, but I don't see a strong concrete difference one way or the other.

Other thing I'd mention is that if you key off the whole stack there's a high potential that you're going to get "the same error" under a number of different stack traces, effectively increasing the number of emails by a (high?) constant factor. 
(There may be many ways to get to offending function F, and they'll each have a different stack trace.) I'm not sure this is going to be a problem but just wanted to raise that if you get way more volume on each single thread than you thought given exponential backoff, keying off just the last N lines of the stacktrace could help.

The other thing I'd add that's very useful to the email is the number of times an error has been triggered. Soft assert does this and that number is a super useful heuristic for how much something is "blowing up"

Comment (benrudolph) [inline corehq/apps/hqwebapp/views.py]: good idea about the stack trace N lines. i'll do that. i also like the blowing up factor. i'll incorporate that as well

**Comment by benrudolph** [comment]
added, but needs this: https://github.com/dimagi/dimagi-utils/pull/271

**Comment by dannyroberts** [comment]
(can't merge)

**Comment by benrudolph** [comment]
updated, i used `theirs` for the submodule commit (that's where the conflict was). it's a bit unclear to me how to check which one is the right one to use

**Comment by benrudolph** [comment]
i think this is good unless you have any more concerns you'd like me to address. i'll take responsibility for for a hotfix or something if the emails are way out of hand

**Pull request merged**